lib/protocol: Avoid data loss on database wipe by higher version numbers (fixes #3876)

calmh opened the pull request:
This makes version vector values clock based instead of just incremented
from zero. The effect is that a vector that is created from scratch
(after database reset) will have a higher value for the local device
than what it could have been previously, causing a conflict. That is, if
we are A and we had

    {A: 42, B: 12}

in the old scheme, a reset and rescan would give us

    {A: 1}

which is a strict ancestor of the older file (this might be wrong). With
the new scheme we would instead have

    {A: someClockTime, b: otherClockTime}

and the new version after reset would become

    {A: someClockTime+delta}

which is in conflict with the previous entry (better).
In case the clocks are wrong (current time is less than the value in the
vector) we fall back to just simple increment like today.

This scheme is ineffective if we suffer a database reset while at the
same time setting the clock back far into the past. It's however no
worse than what we already do.

This loses the ability to emit the "added" event, as we can't look for
the magic 1 entry any more. That event was however already broken
(#5541).

Another place where we infer meaning from the vector itself is in
receive only folders, but there only criteria is that the vector is only
one item long and only includes ourselves, which remain the case with
this change.

AudriusButkevicius commented:
Made a comment about conflict handling. I think we don't do what we used to do, and we will create conflicts even for identical files.

calmh commented:
No, it really doesn't, at least not under good circumstances.

Some file changed multiple times on both sides, then touched, dropped db, restarted on one side;

```
jb@kvin:..thing/syncthing/test % touch s1/syncthing
jb@kvin:..thing/syncthing/test % rm -rf h1/index-v0.14.0.db
jb@kvin:..thing/syncthing/test % STTRACE=scanner,model ../bin/syncthing -home h1
[start] 2020/05/05 14:30:27.659300 main.go:588: INFO: syncthing v1.5.0+36-ga31b412c-dirty-clocks "Fermium Flea" (go1.14.2 darwin-amd64) jb@kvin.kastelo.net 2020-05-05 10:59:05 UTC
...
[I6KAH] 2020/05/05 14:30:29.173467 walk.go:372: DEBUG: to hash: syncthing File{Name:"syncthing", Sequence:0, Permissions:0755, ModTime:2020-05-05 14:30:22.998198 +0200 CEST, Version:{[{I6KAH76 1588681829}]}, Length:24464600, Deleted:false, Invalid:false, LocalFlags:0x0, NoPermissions:false, BlockSize:131072, Blocks:[]}
[I6KAH] 2020/05/05 14:30:29.173496 walk.go:240: DEBUG: ignored (internal): .stfolder
[I6KAH] 2020/05/05 14:30:29.173509 walk.go:193: DEBUG: real to hash: syncthing
[I6KAH] 2020/05/05 14:30:29.173516 folder.go:628: INFO: Completed initial scan of sendreceive folder ¯\_(ツ)_/¯ Räksmörgås 动作 Адрес
...
[I6KAH] 2020/05/05 14:30:29.194056 service.go:326: INFO: Established secure connection to MRIW7OK-NETT3M4-N6SBWME-N25O76W-YJKVXPH-FUMQJ3S-P57B74J-GBITBAC at 127.0.0.1:60878-127.0.0.1:22002/tcp-client/TLS1.3-TLS_AES_128_GCM_SHA256
...
[I6KAH] 2020/05/05 14:30:29.241282 folder.go:628: INFO: Completed initial scan of sendreceive folder default
[I6KAH] 2020/05/05 14:30:29.241303 folder.go:243: DEBUG: sendreceive/default@0xc000196000 next rescan in 11.293150492s
[I6KAH] 2020/05/05 14:30:29.241490 model.go:1969: DEBUG: indexSender@0xc000212070 for default to  at 127.0.0.1:60878-127.0.0.1:22002/tcp-client/TLS1.3-TLS_AES_128_GCM_SHA256: Sending 1 files (<9799 bytes)
[I6KAH] 2020/05/05 14:30:29.255417 model.go:943: DEBUG: Index update (in): MRIW7OK-NETT3M4-N6SBWME-N25O76W-YJKVXPH-FUMQJ3S-P57B74J-GBITBAC / "default": 1 files
[I6KAH] 2020/05/05 14:30:29.255933 folder_sendrecv.go:161: DEBUG: sendreceive/default@0xc000196000 pulling
[I6KAH] 2020/05/05 14:30:29.255985 folder_sendrecv.go:233: DEBUG: sendreceive/default@0xc000196000 copiers: 1 pullerPendingKiB: 32768
[I6KAH] 2020/05/05 14:30:29.256172 folder_sendrecv.go:1140: DEBUG: sendreceive/default@0xc000196000 taking shortcut on syncthing
[I6KAH] 2020/05/05 14:30:29.267717 folder_sendrecv.go:186: DEBUG: sendreceive/default@0xc000196000 changed 1 on try 1
[I6KAH] 2020/05/05 14:30:29.267736 folder_sendrecv.go:233: DEBUG: sendreceive/default@0xc000196000 copiers: 1 pullerPendingKiB: 32768
[I6KAH] 2020/05/05 14:30:29.267806 folder_sendrecv.go:186: DEBUG: sendreceive/default@0xc000196000 changed 0 on try 2
[I6KAH] 2020/05/05 14:30:29.267884 progressemitter.go:296: DEBUG: progress emitter: bytes completed for default: 0
...
```

no conflict files

AudriusButkevicius commented:
I was looking at `ShouldConflict()` which I believe interacts with this in some way, hence the confusion. I am not sure if that still does what it needs.

Going back to the original claim:
```
func (f *sendReceiveFolder) inConflict(current, replacement protocol.Vector) bool {
	if current.Concurrent(replacement) {
		// Obvious case
		return true
	}
	if replacement.Counter(f.shortID) > current.Counter(f.shortID) {
		// The replacement file contains a higher version for ourselves than
		// what we have. This isn't supposed to be possible, since it's only
		// we who can increment that counter. We take it as a sign that
		// something is wrong (our index may have been corrupted or removed)
		// and flag it as a conflict.
		return true
	}
	return false
}
```

This implies that if we see our clock go backwards, we automatically go into conflict, implying the original issue should not be possible to start with?

calmh commented:
Right, that's a thing and it looks promising. It fails in the case where we created a file and someone else deleted it:

- We create: `{A: 1}`
- They delete: `{A: 1, B: 1}`
- We forget
- We recreate: `{A: 1}`

this is an ancestor of the deleted file, nothing strange and no conflict, we just apply the delete again.

(Something else weird happens when the above code actually works and does it's thing, I just tested... It leaves the newly discovered file, it doesn't get synced to the other side, and both consider themselves up to date... I need to research that.)

AudriusButkevicius commented:
Presumably the case you suggested could be addressed by changing > to >= and adding `&& replacement != current`

calmh commented:
I'm not sure why that would be, the `current={A:1}, replacement={A:1, B:1}` case is totally legit however you slice it. It's the same as if we had discovered the file and then later someone deleted it, it's just that the delete entry is already laying about in one of the remote fileinfo sets...

AudriusButkevicius commented:
I guess we should close the "action" type ticket, as it's not something we will fix by the looks of it.
Perhaps we should change the "modified" to "updated"
